doc/manual/src/contributing/experimental-flags.md

```diff
@@ -0,0 +1,68 @@
+This section describes the notion of *experimental features*, and how it fits into the big picture of the development of Nix.
+
+# What are experimental features?
+
+[Since Nix 2.4](../release-notes/rl-2.4.md), some features of Nix can be flagged as experimental.
+Experimental features can be changed or removed at any time, and users must [explicitly enable them].
+This allows accessing new functionality that is not yet stable without unwittingly relying on them.
+
+[explicitly enable them]: ../nix-conf.md#conf-expeirmental-features
+
+# When should a new feature be marked experimental?
+
+A change in the Nix codebase should be guarded by an experimental feature flag if it is likely to either introduce a regression of any kind, or if it implies a non-trivial change to the external interface of Nix.
+
+Examples:
+
+- Changes to the Nix language, such as new built-ins, syntactic or semantic changes, etc.
+- Changes to the command-line interface
+- Some non-trivial internal changes that are likely to introduce bugs
+
+# Lifecycle of an experimental feature
+
+Experimental features have to be treated on a case-by-case basis.
+However, the standard workflow for an experimental feature is as follows:
+
+- The feature is implemented in a pull request.
+- It is merged under an experimental flag, and is disabled by default.
+    - Being experimental, the feature can still be changed arbitrarily or removed again.
+- After at least one release cycle without any significant issues, the feature becomes enabled by default.
```

Comment:
This does not reflect our current process. We don't enable experimental features ever, that's for the user to do.

Reply:
That's true. Then we'd need to specify when to simply make a feature not experimental any more.

Just an idea:

> When maintainers are confident that a feature will not require breaking changes in the future, the experimental flag is removed and the feature considered stable.

Reply:
I like the "cool down" period aspect, which is when we are ready to stabilize something we should make sure it hasn't changed since the last release. That means we might focus on something for stabilization, do some "finishing touches", and then wait a cycle.

Reply:
For that stabilization cycle it is is more of a "feature flag" than an "experimental flag". This would make the feature opt-out vs opt-in.

Reply:
Yeah, calling it “experimental” is wrong here. I think it's a worthwile phase to have for some changes though, where we really want to flip the switch but can't be sure that we won't break something. But maybe we can save it for later